File: site-generation/rss.ml
open General
open Attributes open Css open Nodes open CssConversions open View

open BlogPost open Posts

let generate_rss_for_post (post : blog_post) : string =
    let open String in
    concat [
        "<item><title>";
        post.title;
        "</title><link>https://stars-and-bars.net/posts/";
        post.api_name;
        "</link><description>";
        post.preview;
        "</description><pubDate>";
        of_int post.date.date;
        " ";
        month_to_string post.date.month;
        " ";
        of_int post.date.year;
        "</pubDate><content:encoded><![CDATA[";
        view (blog_post_to_html_node post false);
        "]]></content:encoded></item>"
    ]

let generate_rss (posts : blog_post list) : string =
    let open String in
    concat [
        "<?xml version=\"1.0\" encoding=\"UTF-8\" ?><rss xmlns:content=\"http://purl.org/rss/1.0/modules/content/\" version=\"2.0\"><channel><title>Stars and Bars</title><link>https://stars-and-bars.net/</link><description>A blog about mathematics and computing.</description>";
        posts
        |> List.map (fun x -> generate_rss_for_post x)
        |> List.foldl (fun s m -> s + m) "";
        "</channel></rss>"
    ]

File: site-generation/src/rss.rs
use crate::post::Post;

fn post_to_rss(post : &Post, base_url : &str) -> String {
    let components : Vec<String> = vec![
        "<item><title>".to_string(),
        post.name.to_string(),
        "</title><link>".to_string(),
        base_url.to_string(),
        "posts/".to_string(),
        post.api_name.clone(),
        "</link><description>".to_string(),
        post.preview.clone(),
        "</description><pubDate>".to_string(),
        post.date.day.to_string(),
        " ".to_string(),
        post.date.month.to_string(),
        " ".to_string(),
        post.date.year.to_string(),
        "</pubDate><content:encoded><![CDATA[".to_string(),
        post.html.to_string(),
        "]]></content:encoded></item>".to_string()
    ];

    components.join("")
}

pub fn generate_rss(posts : &Vec<Post>, title : &str, base_url : &str) -> String {

    let mut post_string = String::new();

    for post in posts {
        post_string.push_str(&post_to_rss(post, base_url));
    }

    let components : Vec<String> = vec![
        "<?xml version=\"1.0\" encoding=\"UTF-8\" ?><rss xmlns:content=\"http://purl.org/rss/1.0/modules/content/\" version=\"2.0\"><channel><title>".to_string(),
        title.to_string(),
        "</title><link>".to_string(),
        base_url.to_string(),
        "</link><description>A blog about mathematics and computing.</description>".to_string(),
        post_string,
        "</channel></rss>".to_string()
    ];

    components.join("")
} 


